Src/ILGPU.Algorithms/Random/RandomExtensions.cs

```diff
@@ -153,11 +196,70 @@ internal static ulong ShiftState(ulong state, int laneShift)
         /// <param name="maxNumParallelWarps">
         /// The maximum number of parallel warps.
         /// </param>
+        [CLSCompliant(false)]
         public static RNG<TRandomProvider> CreateRNG<TRandomProvider>(
             this Accelerator accelerator,
             System.Random random,
             int maxNumParallelWarps)
             where TRandomProvider : unmanaged, IRandomProvider<TRandomProvider> =>
             RNG.Create<TRandomProvider>(accelerator, random, maxNumParallelWarps);
+        
+        /// <summary>
+        /// Initialization kernel used to initialize the actual RNG values on the device.
+        /// </summary>
+        internal static void InitializeRNGKernel<TRandomProvider>(
+            Index1D index,
+            ArrayView<XorShift128Plus> sourceProviders,
+            ArrayView<TRandomProvider> randomProvider)
+            where TRandomProvider : unmanaged, IRandomProvider<TRandomProvider>
+        {
+            var provider = sourceProviders[index];
+            for (LongIndex1D i = index;
+                i < randomProvider.Length;
+                i += GridExtensions.GridStrideLoopStride)
+            {
+                randomProvider[i] = default(TRandomProvider).CreateProvider(provider);
+            }
+        }
+        
+        /// <summary>
+        /// Initializes a given view with random values.
+        /// </summary>
+        /// <typeparam name="TRandomProvider">
+        /// The RNG provider type to use on the device.
+        /// </typeparam>
+        /// <param name="accelerator">The current accelerator.</param>
+        /// <param name="stream">The current accelerator stream.</param>
+        /// <param name="rngView">The view to fill.</param>
+        /// <param name="random">The source RNG provider.</param>
+        /// <param name="numInitializers">The number of CPU initializers to use.</param>
+        [CLSCompliant(false)]
+        public static void InitRNGView<TRandomProvider>(
+            this Accelerator accelerator,
+            AcceleratorStream stream,
+            ArrayView<TRandomProvider> rngView,
+            System.Random random,
+            int numInitializers = ushort.MaxValue)
+            where TRandomProvider : unmanaged, IRandomProvider<TRandomProvider>
+        {
+            // Initialize a single provider per warp
+            numInitializers = (int)Math.Min(XMath.Clamp(
+                numInitializers,
+                accelerator.MaxNumThreadsPerGroup,
+                rngView.Length),
+                int.MaxValue >> 4);
```

Comment:
QUESTION: Why shift by 4?

Reply:
I think we should ultimately limit this to `ushort.MaxValue` to avoid large buffers in CPU space.